src/storage/src/source/postgres/replication.rs

```diff
@@ -171,97 +172,98 @@ pub(crate) fn render<G: Scope<Timestamp = MzOffset>>(
+                    // Unfortunately we don't have a way to learn that LSN and so we must choose an
+                    // LSN out of thin air that is guaranteed to not invalidate the definite
+                    // decisions previously made by this operator. We therefore always pick
+                    // `u64::MAX`, which will (in practice) never conflict any previously revealed
+                    // portions of the TVC.
+                    let update = ((oid, Err(err.clone())), MzOffset::from(u64::MAX), 1);
+                    data_output.give(&data_cap_set[0], update).await;
+                }
+                return Ok(());
+            }
 
-            let resume_upper = Antichain::from_iter(
+            // This is an ugly hack that we need to do until because the `build_fallible`
+            // constructor (which we use here) is unsafe in the presence of retained input
+            // capabiltities (which we also use for the rewind requests). The mutable antichain is
+            // initialized with one copy of the minimum capability which represents the initial
+            // capability of this operator.
+            let mut fake_capability = MzOffset::minimum();
+            let mut fake_progress = MutableAntichain::new_bottom(MzOffset::minimum());
+
+            // Calculate the minimum frontier that is required from the slot which is the
+            // combination of the resume upper of each subsource and the requirements of each
+            // rewind request.
+            let mut resume_upper = Antichain::from_iter(
                 subsource_resume_uppers
                     .values()
                     .flat_map(|f| f.elements())
-                    // Advance any upper as far as the slot_lsn.
-                    .map(|t|std::cmp::max(*t, slot_lsn))
+                    .copied()
+                    // If a resume upper of a subsource is zero it means it will be snapshotted and
+                    // the requirement on the overall resume upper will come from its corresponding
+                    // rewind request, so it's safe to ignore it here.
+                    .filter(|t| t != &MzOffset::minimum())
             );
-
-            let Some(resume_lsn) = resume_upper.into_option() else {
-                return Ok(());
-            };
-            data_cap_set.downgrade([&resume_lsn]);
-            upper_cap_set.downgrade([&resume_lsn]);
-            trace!(%id, "timely-{worker_id} replication reader started lsn={}", resume_lsn);
-
-
             let mut rewinds = BTreeMap::new();
             while let Some(event) = rewind_input.next_mut().await {
-                if let AsyncEvent::Data(cap, data) = event {
-                    let cap = cap.retain_for_output(0);
+                if let AsyncEvent::Data(_cap, data) = event {
+                    // We can't retain this capability due to build_fallible so simulate holding it
+                    // by accounting for it in the progress mutable antichain
```

Comment:
more thoroughly explaining _why_ we can hold it would be nice

i think the subtlety is why retaining a data cap is bad if we also have the `data_cap_set` above?